File: backend-spring-boot/src/test/java/no/vicx/backend/ActuatorHealthTest.java
package no.vicx.backend;

import no.vicx.backend.testconfiguration.TestSecurityConfig;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalManagementPort;
import org.springframework.context.annotation.Import;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.test.web.reactive.server.WebTestClient;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsInAnyOrder;
import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureWebTestClient
@Import(TestSecurityConfig.class)
class ActuatorHealthTest {

    @LocalManagementPort
    private int managementPort;

    @Autowired
    WebTestClient webClient;

    @ParameterizedTest
    @MethodSource("probeEndpoints")
    void getHealthProbe_givenExpectedComponents_expectComponentsAsInConfig(
            String probeEndpoint, List<String> expectedComponents) {

        var currentProbeUrl = "http://localhost:%d/actuator/health%s"
                .formatted(managementPort, probeEndpoint);

        webClient.get()
                .uri(currentProbeUrl)
                .exchange()
                .expectStatus().isOk()
                .expectBody(new ParameterizedTypeReference<Map<String, Object>>() {
                })
                .consumeWith(result -> {
                    @SuppressWarnings({"unchecked", "DataFlowIssue"})
                    var actualComponents = (Map<String, Object>) result.getResponseBody().get("components");

                    assertEquals(expectedComponents.size(), actualComponents.size());
                    assertThat(actualComponents.keySet(), containsInAnyOrder(expectedComponents.toArray()));
                });
    }

    private static Stream<Arguments> probeEndpoints() {
        return Stream.of(
                Arguments.of("/readiness", List.of("db", "readinessState")),
                Arguments.of("/liveness", Collections.singletonList("livenessState"))
        );
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/ActuatorHealthTest.kt
package no.vicx.backend

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.data.Row2
import io.kotest.data.forAll
import io.kotest.matchers.collections.shouldContainAll
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.server.LocalManagementPort
import org.springframework.boot.webtestclient.autoconfigure.AutoConfigureWebTestClient
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.test.web.reactive.server.expectBody
import org.springframework.web.util.UriComponentsBuilder

@AutoConfigureWebTestClient
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ActuatorHealthTest(
    @LocalManagementPort managementPort: Int,
    webTestClient: WebTestClient,
) : BehaviorSpec({

        Given("context with health checks") {
            forAll(
                Row2("readiness", setOf("db", "readinessState")),
                Row2("liveness", setOf("livenessState")),
            ) { probeName, expectedComponents ->
                val uri =
                    UriComponentsBuilder
                        .fromUriString("http://localhost:{port}/actuator/health/{probeName}")
                        .buildAndExpand(managementPort, probeName)
                        .toUri()

                Then("$probeName should have $expectedComponents") {
                    webTestClient
                        .get()
                        .uri(uri)
                        .exchange()
                        .expectStatus()
                        .isOk()
                        .expectBody<Map<String, Any>>()
                        .consumeWith { result ->
                            val actualComponents = result.responseBody?.get("components") as Map<*, *>
                            actualComponents.keys shouldContainAll expectedComponents
                        }
                }
            }
        }
    })
